### ISSUE TITLE
after npm install, persistence adapter defaults to node.js 'fs' adapter in web environment

### ISSUE THREAD

**Issue by adueck** [techfort/LokiJS#678]
I am working on a React web-app and I have installed lokijs using npm.

*sidenote: In my current React development setup (using [create-react-app](https://github.com/facebook/create-react-app), which uses webpack etc.), I need to webpack include lokijs from an npm install. I can't get it to work if I just include the regular lokijs.min.js. But maybe that's a seperate issue.*

The problem comes when I try to get things going with persistent storage. Following the [Persistence Adapters Tutorial](https://rawgit.com/techfort/LokiJS/master/jsdoc/tutorial-Persistence%20Adapters.html), my database should default to browser local storage.

However, even in the web environment it seems like it is trying to use the node.js fs adapter. When I try to create the database I get this error:

```
Uncaught TypeError: this.fs.stat is not a function
    at LokiFsAdapter.loadDatabase (lokijs.js:2269)
    at Loki.loadDatabaseInternal (lokijs.js:2461)
    at lokijs.js:2544
    at Loki.throttledSaveDrain (lokijs.js:2440)
    at Loki.loadDatabase (lokijs.js:2539)
    at lokijs.js:951
```

Here is the code I am using to create the persisting database:

```
var lokidb = new loki("localDB.db", {
  autoload: true,
  autoloadCallback : databaseInitialize,
  autosave: true,
  autosaveInterval: 10000
});

function databaseInitialize() {
  var localDB = lokidb.getCollection("localDB");

  if (localDB === null) {
    localDB = lokidb.addCollection("localDB");
  }
}
```

3 questions about this: 

- Shouldn't the npm package detect whether it's running in a node.js or a web environment? 
- If not, what's the adapter property we can pass to specify web local storage? I guessed and tried using `adapter: localStorageAdapter` but that doesn't seem to work.
- Also, If we want to use the idb adapter, is there an npm installable, webpack importable version of `loki-indexed-adapter.js`?

Thank you so much for all your great work on this.

**Comment by obeliskos** [comment]
That's odd... so you can reproduce this in chrome on your dev machine?  The environment detection is within the constructor and currently looks like this :

https://github.com/techfort/LokiJS/blob/46a341184180f57cfa87b8d14691e2932d18efe9/src/lokijs.js#L828#L849

You could put a breakpoint in that inline function and see why it is determining you are using node.
As a workaround, you might pass in the environment in call to constructor, such as : 

```javascript
var lokidb = new loki("localDB.db", {
  env: 'BROWSER',
  autoload: true,
  autoloadCallback : databaseInitialize,
  autosave: true,
  autosaveInterval: 10000
});

```

As for a proper indexeddb module you can import/require, you -might- be able to use our LokiDB indexed db adapter.  While the project is in beta, that adapter (at least in its current version) should be identical to our bundled lokijs indexedb adapter.  (@Viatorus can clarify if otherwise).  You can find a link to that [indexed-storage adapter here](https://www.npmjs.com/package/@lokidb/loki) or try installing with : 
```
npm install @lokidb/indexed-storage
```

**Comment by adueck** [comment]
Thank you for your very quick response!

The workaround with `env: 'BROWSER'` works perfectly. Thank you.

As for the cause of the problem, I added a breakpoint and saw that in Chrome Ver. 66.0.3359.139 on Windows 10 it is indeed returning 'NODEJS' on line 839.

`(typeof global !== 'undefined' && global.window && process)` evaluates as true and so it executes line 839:

`return 'NODEJS'; //node-webkit`

The exact same thing happens in Firefox Ver. 60.0.1

*Interestingly, I also tried using the beta lokiDB with `npm install @lokidb/loki` which had a different a different problem. It didn't detect a node environment, but it wouldn't call the databaseInitialize() function specified in `autoLoadCallback`. I imagine this is just a difference in the API for the new lokiDB that will be explained when the documentation comes out.*

I'm happy to try and see if the lokiDB indexed-storage adapter works. I did `npm install @loki/indexed-storage`. But then how would I import it? 

```
import loki from 'lokijs';
import ????? from '@lokidb/indexed-storage';
```

Thanks for your patience, I'm still a beginner in JavaScript!

**Comment by obeliskos** [comment]
Not sure why global or process would be set...  

Not sure but the export is called IndexedStorage, so maybe... ? : 
```
import { IndexedStorage } from '@lokidb/indexed-storage';
```

**Comment by adueck** [comment]
> Not sure why global or process would be set...

Could it be because I'm running a development server with create-react-app on localhost:3000 ?? 

> Not sure but the export is called IndexedStorage, so maybe... ? :
> 
> import { IndexedStorage } from '@lokidb/indexed-storage';

Unfortunately this doesn't quite work either...

```
import loki from `lokijs`;
import { IndexedStorage } from '@lokidb/indexed-storage';

...

var idbAdapter = new LokiIndexedAdapter();
var lokidbPersistant = new loki("suggestions.db", {
  env: `BROWSER`,
  adapter: idbAdapter,
  autoload: true,
  autoloadCallback : databaseInitialize,
  autosave: true,
  autosaveInterval: 10000
});
```

returns an error: `'LokiIndexedAdapter' is not defined  no-undef` of course...

and when I try to do this... (wild stab in the dark)

```
import loki from `lokijs`;
import { IndexedStorage } from '@lokidb/indexed-storage';

...

var idbAdapter = new IndexedStorage.LokiIndexedAdapter();
var lokidbPersistant = new loki("suggestions.db", {
  env: `BROWSER`,
  adapter: idbAdapter,
  autoload: true,
  autoloadCallback : databaseInitialize,
  autosave: true,
  autosaveInterval: 10000
});
```

It doesn't work either: `IndexedStorage.LokiIndexedAdapter is not a constructor`

Also, there seems to be some issue with that package, because when I try to compile and minify it all with webpack I get this error:

```
Failed to minify the code from this file:

        ./node_modules/@lokidb/indexed-storage/lokidb.indexed-storage.js:103
```

**Comment by Viatorus** [comment]
Just a note from my side:

You are mixing LokiJS with LokiDB which is not supported and maybe not
probably work.

There is also a IndexedStorageAdapter in this repository for LokiJS.

The one from LokiDB uses ES6 syntax so that's why your minifying does not
work.

>

**Comment by adueck** [comment]
Yep, sorry I realized I was getting a bit off topic for this thread.

> You are mixing LokiJS with LokiDB which is not supported and maybe not probably work. 
> There is also a IndexedStorageAdapter in this repository for LokiJS. 

Yep there is, but I was having trouble including it with webpack, but I think I figured out a way to do it. 👍 

>The one from LokiDB uses ES6 syntax so that's why your minifying does not work.

Hmm... I'm also minifying lots of stuff with ES6 syntax so I'm not sure that's the problem, so might be worth looking into. 

Keep up the fantastic work guys, overall very thankful for this stuff!
